From 0db52786f7c7977d337e8b823a8fc2952a34d8dd Mon Sep 17 00:00:00 2001
From: Zhang Zekun <zhangzekun11@huawei.com>
Date: Mon, 14 Nov 2022 11:21:19 +0000
Subject: [PATCH 0063/1398] [Huawei] vmalloc: Add config for Extend for
 hugepages mapping

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187974

----------------------------------------

Use CONFIG_EXTEND_HUGEPAGE_MAPPING to isolate code
introduced in a3425d4173784e41644abc3cc70d41aa2aef16fd.

Besides, use tab instead of space to match the format
of Kconfig

Signed-off-by: Zhang Zekun <zhangzekun11@huawei.com>
---
 arch/arm64/Kconfig                     |  2 +-
 arch/arm64/configs/openeuler_defconfig |  1 +
 include/linux/mm.h                     |  2 ++
 include/linux/vmalloc.h                |  7 +++++++
 mm/Kconfig                             | 17 ++++++++++++-----
 mm/vmalloc.c                           | 16 ++++++++++++++++
 6 files changed, 39 insertions(+), 6 deletions(-)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 6c249ee0fe12..8776690e1a7c 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -2109,7 +2109,7 @@ config ASCEND_CHARGE_MIGRATE_HUGEPAGES
 config ASCEND_SHARE_POOL
 	bool "Enable support for the Share Pool Memory"
 	default n
-	depends on HAVE_ARCH_HUGE_VMALLOC
+	depends on HAVE_ARCH_HUGE_VMALLOC && EXTEND_HUGEPAGE_MAPPING
 	select ARCH_USES_HIGH_VMA_FLAGS
 	help
 	  This feature allows multiple processes to share virtual memory both
diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index 3544ea64b79d..193133b67149 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -1125,6 +1125,7 @@ CONFIG_PIN_MEMORY=y
 CONFIG_PID_RESERVE=y
 CONFIG_MEMORY_RELIABLE=y
 CONFIG_EXPORT_COLLECT_PROCS=y
+CONFIG_EXTEND_HUGEPAGE_MAPPING=y
 
 #
 # Data Access Monitoring
diff --git a/include/linux/mm.h b/include/linux/mm.h
index 2f97855321e7..7f0ca7ce700a 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -234,11 +234,13 @@ int overcommit_policy_handler(struct ctl_table *, int, void *, size_t *,
 /* test whether an address (unsigned long or pointer) is aligned to PAGE_SIZE */
 #define PAGE_ALIGNED(addr)	IS_ALIGNED((unsigned long)(addr), PAGE_SIZE)
 
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 /* to align the pointer to the (next) PMD hugepage boundary */
 #define PMD_ALIGN(addr) ALIGN(addr, PMD_SIZE)
 
 /* test whether an address (unsigned long or pointer) is aligned to PMD_SIZE */
 #define PMD_ALIGNED(addr)   IS_ALIGNED((unsigned long)(addr), PMD_SIZE)
+#endif
 
 #define lru_to_page(head) (list_entry((head)->prev, struct page, lru))
 
diff --git a/include/linux/vmalloc.h b/include/linux/vmalloc.h
index c37ea39fd7b6..570d29966b56 100644
--- a/include/linux/vmalloc.h
+++ b/include/linux/vmalloc.h
@@ -27,7 +27,9 @@ struct notifier_block;		/* in notifier.h */
 #define VM_FLUSH_RESET_PERMS	0x00000100	/* reset direct map and flush TLB on unmap, can't be freed in atomic context */
 #define VM_MAP_PUT_PAGES	0x00000200	/* put pages and free array in vfree */
 #define VM_NO_HUGE_VMAP		0x00000400	/* force PAGE_SIZE pte mapping */
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 #define VM_HUGE_PAGES		0x00001000	/* used for vmalloc hugepages */
+#endif
 #ifdef CONFIG_ASCEND_SHARE_POOL
 #define VM_SHAREPOOL		0x00002000      /* remapped to sharepool */
 #else
@@ -148,8 +150,11 @@ extern void *__vmalloc_node_range(unsigned long size, unsigned long align,
 void *__vmalloc_node(unsigned long size, unsigned long align, gfp_t gfp_mask,
 		int node, const void *caller);
 void *vmalloc_no_huge(unsigned long size);
+
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 extern void *vmalloc_hugepage(unsigned long size);
 extern void *vmalloc_hugepage_user(unsigned long size);
+#endif
 
 extern void vfree(const void *addr);
 extern void vfree_atomic(const void *addr);
@@ -166,6 +171,7 @@ extern int remap_vmalloc_range_partial(struct vm_area_struct *vma,
 extern int remap_vmalloc_range(struct vm_area_struct *vma, void *addr,
 							unsigned long pgoff);
 
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 extern void *vmap_hugepage(struct page **pages, unsigned int count,
 			   unsigned long flags, pgprot_t prot);
 extern int remap_vmalloc_hugepage_range_partial(struct vm_area_struct *vma,
@@ -173,6 +179,7 @@ extern int remap_vmalloc_hugepage_range_partial(struct vm_area_struct *vma,
 						unsigned long pgoff, unsigned long size);
 extern int remap_vmalloc_hugepage_range(struct vm_area_struct *vma,
 					void *addr, unsigned long pgoff);
+#endif
 
 /*
  * Architectures can set this mask to a combination of PGTBL_P?D_MODIFIED values
diff --git a/mm/Kconfig b/mm/Kconfig
index eef871ccf465..4abc922fd669 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -1018,11 +1018,18 @@ config CLEAR_FREELIST_PAGE
 	  needs to be added.
 
 config EXPORT_COLLECT_PROCS
-       bool "Export the function collect_procs()"
-       default n
-       help
-         Export this function to collect the processes who have the page
-         mapped via collect_procs().
+	bool "Export the function collect_procs()"
+	default n
+	help
+	  Export this function to collect the processes who have the page
+	  mapped via collect_procs().
+
+config EXTEND_HUGEPAGE_MAPPING
+	bool "Extend for hugepages mapping"
+	depends on ARM64
+	default n
+	help
+	  Introduce vmalloc/vmap/remap interfaces that handle only hugepages.
 
 source "mm/damon/Kconfig"
 
diff --git a/mm/vmalloc.c b/mm/vmalloc.c
index 1769dc4d46a2..19474c3378aa 100644
--- a/mm/vmalloc.c
+++ b/mm/vmalloc.c
@@ -578,6 +578,7 @@ static int vmap_pages_range(unsigned long addr, unsigned long end,
 	return err;
 }
 
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 static int vmap_hugepages_range_noflush(unsigned long addr, unsigned long end,
 		pgprot_t prot, struct page **pages, unsigned int page_shift)
 {
@@ -609,6 +610,7 @@ static int vmap_hugepages_range(unsigned long addr, unsigned long end,
 
 	return err;
 }
+#endif
 
 /**
  * map_kernel_range_noflush - map kernel VM area with the specified pages
@@ -2792,6 +2794,7 @@ void *vmap(struct page **pages, unsigned int count,
 }
 EXPORT_SYMBOL(vmap);
 
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 /**
  *	vmap_hugepage  -  map an array of huge pages into virtually contiguous space
  *	@pages:		array of huge page pointers (only the header)
@@ -2830,6 +2833,7 @@ void *vmap_hugepage(struct page **pages, unsigned int count,
 	return area->addr;
 }
 EXPORT_SYMBOL(vmap_hugepage);
+#endif
 
 #ifdef CONFIG_VMAP_PFN
 struct vmap_pfn_data {
@@ -3015,7 +3019,11 @@ void *__vmalloc_node_range(unsigned long size, unsigned long align,
 		size_per_node = size;
 		if (node == NUMA_NO_NODE)
 			size_per_node /= num_online_nodes();
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 		if (size_per_node >= PMD_SIZE || vm_flags & VM_HUGE_PAGES) {
+#else
+		if (size_per_node >= PMD_SIZE) {
+#endif
 			shift = PMD_SHIFT;
 			align = max(real_align, 1UL << shift);
 			size = ALIGN(real_size, 1UL << shift);
@@ -3051,8 +3059,12 @@ void *__vmalloc_node_range(unsigned long size, unsigned long align,
 	return addr;
 
 fail:
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 	/* User could specify VM_HUGE_PAGES to alloc only hugepages. */
 	if (shift > PAGE_SHIFT && !(vm_flags & VM_HUGE_PAGES)) {
+#else
+	if (shift > PAGE_SHIFT) {
+#endif
 		shift = PAGE_SHIFT;
 		align = real_align;
 		size = real_size;
@@ -3262,6 +3274,7 @@ void *vmalloc_32_user(unsigned long size)
 }
 EXPORT_SYMBOL(vmalloc_32_user);
 
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 /**
  * vmalloc_hugepage - allocate virtually contiguous hugetlb memory
  *	@size:          allocation size
@@ -3299,6 +3312,7 @@ void *vmalloc_hugepage_user(unsigned long size)
 				    __builtin_return_address(0));
 }
 EXPORT_SYMBOL(vmalloc_hugepage_user);
+#endif
 
 /*
  * small helper routine , copy contents to buf from addr.
@@ -3621,6 +3635,7 @@ int remap_vmalloc_range(struct vm_area_struct *vma, void *addr,
 }
 EXPORT_SYMBOL(remap_vmalloc_range);
 
+#ifdef CONFIG_EXTEND_HUGEPAGE_MAPPING
 /**
  *	remap_vmalloc_hugepage_range_partial - map vmalloc hugepages
  *	to userspace
@@ -3707,6 +3722,7 @@ int remap_vmalloc_hugepage_range(struct vm_area_struct *vma, void *addr,
 						    vma->vm_end - vma->vm_start);
 }
 EXPORT_SYMBOL(remap_vmalloc_hugepage_range);
+#endif
 
 void free_vm_area(struct vm_struct *area)
 {
-- 
2.17.1

